File: app/src/main/java/top/asdb/codexremote/codex/ThreadSessionCache.kt
package top.asdb.codexremote.codex

import top.asdb.codexremote.data.CodexThread
import top.asdb.codexremote.data.TimelineEntry
import top.asdb.codexremote.data.TokenUsage
import top.asdb.codexremote.data.hasKnownContextWindow
import java.util.LinkedHashMap

/** In-memory, per-server LRU cache used as a fast UI/timeout fallback for opened threads. */
class ThreadSessionCache(
    private val maxEntries: Int = DEFAULT_MAX_ENTRIES,
    private val ttlMs: Long = DEFAULT_TTL_MS,
    private val nowMs: () -> Long = System::currentTimeMillis,
    private val maxWeightChars: Int = DEFAULT_MAX_WEIGHT_CHARS,
) {
    init {
        require(maxEntries > 0) { "maxEntries must be positive" }
        require(ttlMs > 0) { "ttlMs must be positive" }
        require(maxWeightChars > 0) { "maxWeightChars must be positive" }
    }

    data class Snapshot(
        val thread: CodexThread,
        val timeline: List<TimelineEntry>,
        val savedAtMs: Long,
        val nextTurnsCursor: String? = null,
        /** Latest server-reported usage for this thread's current context window. */
        val tokenUsage: TokenUsage? = null,
    )

    private data class WeightedSnapshot(val value: Snapshot, val weightChars: Int)

    private val entries = object : LinkedHashMap<String, WeightedSnapshot>(maxEntries, 0.75f, true) {}
    // Context usage is tiny and must survive an oversized transcript being rejected from the
    // display cache. It remains session-scoped; a fresh server notification is authoritative.
    private val contextUsages = object : LinkedHashMap<String, TokenUsage>(maxEntries, 0.75f, true) {}
    private var currentWeightChars = 0

    @Synchronized
    fun get(threadId: String): Snapshot? {
        val snapshot = entries[threadId]?.value ?: return null
        if (nowMs() - snapshot.savedAtMs > ttlMs) {
            return null
        }
        return snapshot.withLatestContextUsage(threadId)
    }

    /** Returns an expired snapshot too, for a last-resort timeout fallback. */
    @Synchronized
    fun getStale(threadId: String): Snapshot? = entries[threadId]?.value?.withLatestContextUsage(threadId)

    /** Returns the latest usable context value even when no transcript snapshot was retained. */
    @Synchronized
    fun contextUsage(threadId: String): TokenUsage? = contextUsages[threadId]

    @Synchronized
    fun put(
        thread: CodexThread,
        timeline: List<TimelineEntry>,
        nextTurnsCursor: String? = null,
        tokenUsage: TokenUsage? = null,
    ) {
        tokenUsage?.takeIf { it.hasKnownContextWindow() }?.let { usage ->
            contextUsages[thread.id] = usage
            while (contextUsages.size > maxContextUsageEntries) {
                contextUsages.remove(contextUsages.entries.first().key)
            }
        }
        val rememberedUsage = contextUsages[thread.id]
        val weight = snapshotWeight(thread, timeline, nextTurnsCursor)
        if (weight > maxWeightChars) return
        removeLocked(thread.id)
        val snapshot = Snapshot(thread, timeline.toList(), nowMs(), nextTurnsCursor, rememberedUsage)
        entries[thread.id] = WeightedSnapshot(snapshot, weight)
        currentWeightChars += weight
        while (entries.size > maxEntries || currentWeightChars > maxWeightChars) {
            removeLocked(entries.entries.first().key)
        }
    }

    @Synchronized
    fun remove(threadId: String) {
        removeLocked(threadId)
        contextUsages.remove(threadId)
    }

    @Synchronized
    fun clear() {
        entries.clear()
        contextUsages.clear()
        currentWeightChars = 0
    }

    @Synchronized
    fun size(): Int = entries.size

    private fun removeLocked(threadId: String) {
        entries.remove(threadId)?.let { currentWeightChars -= it.weightChars }
    }

    private fun Snapshot.withLatestContextUsage(threadId: String): Snapshot = copy(
        tokenUsage = contextUsages[threadId] ?: tokenUsage,
    )

    private fun snapshotWeight(
        thread: CodexThread,
        timeline: List<TimelineEntry>,
        nextTurnsCursor: String?,
    ): Int {
        var result = thread.id.length + thread.title.length + thread.preview.length + thread.cwd.length +
            thread.source.length + thread.status.length + thread.cliVersion.length + nextTurnsCursor.orEmpty().length
        result = saturatedAdd(result, estimateTimelineWeightChars(timeline))
        return result
    }

    private fun saturatedAdd(left: Int, right: Int): Int =
        if (right > Int.MAX_VALUE - left) Int.MAX_VALUE else left + right

    companion object {
        const val DEFAULT_MAX_ENTRIES = 8
        const val DEFAULT_TTL_MS = 30 * 60 * 1000L
        const val DEFAULT_MAX_WEIGHT_CHARS = 2 * 1024 * 1024
        private const val CONTEXT_USAGE_ENTRIES_PER_TRANSCRIPT = 4
    }

    private val maxContextUsageEntries: Int =
        maxEntries.coerceAtMost(Int.MAX_VALUE / CONTEXT_USAGE_ENTRIES_PER_TRANSCRIPT) *
            CONTEXT_USAGE_ENTRIES_PER_TRANSCRIPT
}

internal fun estimateTimelineWeightChars(timeline: List<TimelineEntry>): Int {
    var result = 0
    timeline.forEach { entry ->
        result = saturatedAdd(
            result,
            entry.id.length + entry.title.length + entry.text.length + entry.status.length +
                entry.command.length + entry.cwd.length + entry.output.length + entry.turnId.length +
                entry.subAgentPath.length + entry.subAgentThreadId.length + entry.subAgentActivity.length,
        )
        entry.reasoningSummary.forEach { result = saturatedAdd(result, it.length) }
        entry.reasoningContent.forEach { result = saturatedAdd(result, it.length) }
        entry.changes.forEach { change ->
            result = saturatedAdd(result, change.path.length + change.kind.length + change.diff.length)
        }
    }
    return result
}

private fun saturatedAdd(left: Int, right: Int): Int =
    if (right > Int.MAX_VALUE - left) Int.MAX_VALUE else left + right

File: flutter_app/lib/src/agent/thread_session_cache.dart
import 'dart:collection';

import '../domain/models.dart';

class ThreadSessionSnapshot {
  ThreadSessionSnapshot({
    required this.thread,
    required List<TimelineEntry> timeline,
    required this.savedAtEpochMillis,
    this.nextTurnsCursor,
    this.tokenUsage,
  }) : timeline = List.unmodifiable(timeline);

  final AgentThread thread;
  final List<TimelineEntry> timeline;
  final int savedAtEpochMillis;
  final String? nextTurnsCursor;
  final TokenUsage? tokenUsage;

  ThreadSessionSnapshot copyWith({TokenUsage? tokenUsage}) =>
      ThreadSessionSnapshot(
        thread: thread,
        timeline: timeline,
        savedAtEpochMillis: savedAtEpochMillis,
        nextTurnsCursor: nextTurnsCursor,
        tokenUsage: tokenUsage ?? this.tokenUsage,
      );
}

/// Per-server in-memory cache used to render a resumed thread immediately.
class ThreadSessionCache {
  ThreadSessionCache({
    this.maxEntries = defaultMaxEntries,
    this.ttl = defaultTtl,
    this.maxWeightChars = defaultMaxWeightChars,
    int Function()? nowEpochMillis,
  }) : _nowEpochMillis =
           nowEpochMillis ?? (() => DateTime.now().millisecondsSinceEpoch) {
    if (maxEntries < 1) throw ArgumentError.value(maxEntries, 'maxEntries');
    if (ttl <= Duration.zero) throw ArgumentError.value(ttl, 'ttl');
    if (maxWeightChars < 1) {
      throw ArgumentError.value(maxWeightChars, 'maxWeightChars');
    }
  }

  static const defaultMaxEntries = 8;
  static const defaultTtl = Duration(minutes: 30);
  static const defaultMaxWeightChars = 2 * 1024 * 1024;
  static const _contextUsageEntriesPerTranscript = 4;

  final int maxEntries;
  final Duration ttl;
  final int maxWeightChars;
  final int Function() _nowEpochMillis;
  final LinkedHashMap<String, _WeightedSnapshot> _entries = LinkedHashMap();
  final LinkedHashMap<String, TokenUsage> _contextUsages = LinkedHashMap();
  int _currentWeightChars = 0;

  int get length => _entries.length;

  ThreadSessionSnapshot? get(String threadId) {
    final weighted = _touch(_entries, threadId);
    final snapshot = weighted?.snapshot;
    if (snapshot == null) return null;
    if (_nowEpochMillis() - snapshot.savedAtEpochMillis > ttl.inMilliseconds) {
      return null;
    }
    return _withLatestContextUsage(snapshot, threadId);
  }

  /// Returns an expired snapshot too, for a timeout fallback while reconnecting.
  ThreadSessionSnapshot? getStale(String threadId) {
    final snapshot = _touch(_entries, threadId)?.snapshot;
    return snapshot == null
        ? null
        : _withLatestContextUsage(snapshot, threadId);
  }

  TokenUsage? contextUsage(String threadId) => _touch(_contextUsages, threadId);

  void put(
    AgentThread thread,
    List<TimelineEntry> timeline, {
    String? nextTurnsCursor,
    TokenUsage? tokenUsage,
  }) {
    if (tokenUsage?.hasKnownContextWindow ?? false) {
      _contextUsages.remove(thread.id);
      _contextUsages[thread.id] = tokenUsage!;
      while (_contextUsages.length > _maxContextUsageEntries) {
        _contextUsages.remove(_contextUsages.keys.first);
      }
    }

    final weight = _snapshotWeight(thread, timeline, nextTurnsCursor);
    if (weight > maxWeightChars) return;

    _removeSnapshot(thread.id);
    final snapshot = ThreadSessionSnapshot(
      thread: thread,
      timeline: timeline,
      savedAtEpochMillis: _nowEpochMillis(),
      nextTurnsCursor: nextTurnsCursor,
      tokenUsage: _touch(_contextUsages, thread.id),
    );
    _entries[thread.id] = _WeightedSnapshot(snapshot, weight);
    _currentWeightChars += weight;
    while (_entries.length > maxEntries ||
        _currentWeightChars > maxWeightChars) {
      _removeSnapshot(_entries.keys.first);
    }
  }

  void remove(String threadId) {
    _removeSnapshot(threadId);
    _contextUsages.remove(threadId);
  }

  /// Marks a completed child thread in every retained parent transcript.
  ///
  /// Child turns are delivered on their own thread, so their terminal event
  /// would otherwise wait for the parent turn to end before its cached
  /// collaborator rows are refreshed.
  bool updateSubAgentStatus(String subAgentThreadId, String status) {
    final childId = subAgentThreadId.trim();
    final nextStatus = status.trim();
    if (childId.isEmpty || nextStatus.isEmpty || _entries.isEmpty) return false;

    var changed = false;
    for (final entry in _entries.entries.toList(growable: false)) {
      final snapshot = entry.value.snapshot;
      var timelineChanged = false;
      final timeline = snapshot.timeline
          .map((item) {
            if (item.kind != TimelineKind.subAgent ||
                item.subAgentThreadId != childId ||
                !_isActiveSubAgentStatus(item.status) ||
                item.status == nextStatus) {
              return item;
            }
            timelineChanged = true;
            return item.copyWith(status: nextStatus);
          })
          .toList(growable: false);
      if (!timelineChanged) continue;

      final replacement = ThreadSessionSnapshot(
        thread: snapshot.thread,
        timeline: timeline,
        savedAtEpochMillis: snapshot.savedAtEpochMillis,
        nextTurnsCursor: snapshot.nextTurnsCursor,
        tokenUsage: snapshot.tokenUsage,
      );
      final nextWeight = _snapshotWeight(
        replacement.thread,
        replacement.timeline,
        replacement.nextTurnsCursor,
      );
      _entries[entry.key] = _WeightedSnapshot(replacement, nextWeight);
      _currentWeightChars += nextWeight - entry.value.weightChars;
      changed = true;
    }
    while (_entries.length > maxEntries ||
        _currentWeightChars > maxWeightChars) {
      _removeSnapshot(_entries.keys.first);
    }
    return changed;
  }

  void clear() {
    _entries.clear();
    _contextUsages.clear();
    _currentWeightChars = 0;
  }

  int get _maxContextUsageEntries =>
      maxEntries * _contextUsageEntriesPerTranscript;

  ThreadSessionSnapshot _withLatestContextUsage(
    ThreadSessionSnapshot snapshot,
    String threadId,
  ) {
    final latest = _touch(_contextUsages, threadId);
    return latest == null ? snapshot : snapshot.copyWith(tokenUsage: latest);
  }

  void _removeSnapshot(String threadId) {
    final removed = _entries.remove(threadId);
    if (removed != null) _currentWeightChars -= removed.weightChars;
  }

  int _snapshotWeight(
    AgentThread thread,
    List<TimelineEntry> timeline,
    String? nextTurnsCursor,
  ) {
    var result =
        thread.id.length +
        thread.title.length +
        thread.preview.length +
        thread.cwd.length +
        thread.source.length +
        thread.status.length +
        thread.cliVersion.length +
        (nextTurnsCursor?.length ?? 0);
    result += estimateTimelineWeightChars(timeline);
    return result;
  }
}

class _WeightedSnapshot {
  const _WeightedSnapshot(this.snapshot, this.weightChars);

  final ThreadSessionSnapshot snapshot;
  final int weightChars;
}

V? _touch<V>(LinkedHashMap<String, V> values, String key) {
  final value = values.remove(key);
  if (value != null) values[key] = value;
  return value;
}

bool _isActiveSubAgentStatus(String status) => const <String>{
  'pendingInit',
  'running',
  'inProgress',
  'started',
  'interacted',
  'unknown',
}.contains(status);

int estimateTimelineWeightChars(List<TimelineEntry> timeline) {
  var result = 0;
  for (final entry in timeline) {
    result +=
        entry.id.length +
        entry.title.length +
        entry.text.length +
        entry.status.length +
        entry.command.length +
        entry.cwd.length +
        entry.output.length +
        entry.turnId.length +
        entry.subAgentPath.length +
        entry.subAgentThreadId.length +
        entry.subAgentActivity.length;
    for (final value in entry.reasoningSummary) {
      result += value.length;
    }
    for (final value in entry.reasoningContent) {
      result += value.length;
    }
    for (final change in entry.changes) {
      result += change.path.length + change.kind.length + change.diff.length;
    }
    for (final attachment in entry.attachments) {
      result +=
          attachment.name.length +
          attachment.remotePath.length +
          attachment.mimeType.length;
    }
  }
  return result;
}
